src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -381,10 +393,8 @@ private async Task RefreshKeyValueCollections()
         {
             foreach (KeyValueWatcher changeWatcher in _options.MultiKeyWatchers)
             {
-                var timeElapsedSinceLastRefresh = DateTimeOffset.UtcNow - changeWatcher.LastRefreshTime;
-
                 // Skip the refresh for this key-prefix if the cached value has not expired or a refresh operation is in progress
-                if (timeElapsedSinceLastRefresh < changeWatcher.CacheExpirationTime || !changeWatcher.Semaphore.Wait(0))
+                if (DateTimeOffset.UtcNow < changeWatcher.CacheExpirationTime || !changeWatcher.Semaphore.Wait(0))

```

Comment:
We should check against same point in time.

Reply:
If the current time equals cache expiration time, we do want to refresh, so this condition is correct.

---
In reply to: [438453710](https://github.com/Azure/AppConfiguration-DotnetProvider/pull/165#discussion_r438453710) [](ancestors = 438453710)

Reply:
That's my point. Take a snapshot of Now and evaluate based on it, instead of moving Now in each iteration.

Reply:
If user invokes RefreshAsync at 12:00:00, first registered key-label has cache expiration time 11:59:59 and second key-label has cache expiration time 12:00:02, then taking a snapshot at 12:00:00 would not refresh the second key-label even when we take 60 seconds to refresh the first key-value. Checking the current time enables us to fetch the updated value in case it is stale by the time we reach it. The goal is to refresh as often as possible without violating the minimum cache expiration interval.

Reply:
Make sense. Wanted to make sure it's by design.